crates/indexer/src/users.rs

```diff
@@ -0,0 +1,60 @@
+use anyhow::{anyhow, Result};
+
+use async_trait::async_trait;
+
+use ring::rand::{SecureRandom, SystemRandom};
+
+use crate::iface::{Flush, UserMapper};
+
+const REGISTERED_USERS_MAP: &str = "registered_users";
+
+pub struct UsersStore {
+    map: sled::Tree,
+    rng: SystemRandom,
+}
+
+impl UsersStore {
+    pub fn new(db: &sled::Db) -> Result<Self> {
+        let map = db.open_tree(REGISTERED_USERS_MAP)?;
+
+        let rng = SystemRandom::new();
+
+        Ok(Self { map, rng })
+    }
+
+    fn generate_salt(&self) -> Result<[u8; 16]> {
+        let mut salt = [0u8; 16];
+        self.rng.fill(&mut salt).map_err(|e| anyhow!("{}", e))?;
+        Ok(salt)
+    }
+}
+
+#[async_trait]
+impl Flush for UsersStore {
+    async fn flush(&self) -> Result<()> {
+        self.map.flush_async().await?;
+        Ok(())
+    }
+}
+
+impl UserMapper for UsersStore {
+    fn add_user(&self, username: &str, password: &str) -> Result<()> {
```

Comment:
I might be missing something, but I don't think there's any check asserting that the username wasn't already taken.

Reply:
Right - on it